### ISSUE TITLE
Checkpointing in Optim

### ISSUE THREAD

**Issue by umbriquse** [JuliaNLSolvers/Optim.jl#884]
I've come across the scenario of using an optimization algorithm, but did not have the time in the rest of the day to let the optimizer complete and find a solution. I was curious if you thought about putting a checkpointing system in Optim to allow users to stop their algorithms at an arbitrary point and continue off at that point at another time (i.e. Save the progress that the optimizer has already made)

**Comment by pkofod** [comment]
How would you stop it? With SIGINT ?

**Comment by umbriquse** [comment]
Yes, but also in the situation of either power failure, an error occurring in the user code after some iterations, or some segfault/ bug in julia that hasn't been corrected. Mostly with SIGINT though.

**Comment by umbriquse** [comment]
This idea comes from another package that I am using called [DFTK](https://github.com/JuliaMolSim/DFTK.jl). They rely on using [JLD2](https://github.com/JuliaIO/JLD2.jl), and save a snapshot of the current state so that the progress made isn't lost and can be continued at another time. If you're curious it's in the [jld2io.jl](https://github.com/JuliaMolSim/DFTK.jl/blob/master/src/jld2io.jl) folder.

**Comment by pkofod** [comment]
Okay okay, I think I may understand. If you're that worried about sudden problems I would suggest you use a callback to save the current state. Does that make sense?

**Comment by umbriquse** [comment]
Yes, I thought it wasn't possible to save the progress of the state and then start an optimizer in at that saved state when using any callback function. Thank you, I'll do that.

**Comment by pkofod** [comment]
It's not super easy to start it again, but I can help you set it up if you need something like this.

**Comment by umbriquse** [comment]
Thank you for offering, but I think I can work it out. I am curious if you would like the function that I would make though? So others could add that function into the callback and save the state of their optimizers.

**Comment by pkofod** [comment]
Yeah, please show it. If nothing else we can put it in the docs :)
